eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxFunder.scala

```diff
@@ -405,75 +353,46 @@ private class ReplaceableTxFunder(nodeParams: NodeParams,
     // Since the purpose of this transaction is just to do a CPFP, the resulting tx should have a single change output
-    // (note that bitcoind doesn't let us publish a transaction with no outputs).
-    // To work around these limitations, we start with a dummy output and later merge that dummy output with the optional
-    // change output added by bitcoind.
-    // NB: fundrawtransaction doesn't support non-wallet inputs, so we have to remove our anchor input and re-add it later.
-    // That means bitcoind will not take our anchor input's weight into account when adding inputs to set the fee.
-    // That's ok, we can increase the fee later by decreasing the output amount. But we need to ensure we'll have enough
-    // to cover the weight of our anchor input, which is why we set it to the following value.
-    val dummyChangeAmount = weight2fee(anchorFeerate, claimAnchorOutputMinWeight) + dustLimit
-    val txNotFunded = Transaction(2, Nil, TxOut(dummyChangeAmount, Script.pay2wpkh(PlaceHolderPubKey)) :: Nil, 0)
-    bitcoinClient.fundTransaction(txNotFunded, FundTransactionOptions(anchorFeerate, lockUtxos = true)).flatMap(fundTxResponse => {
+    // (note that bitcoind doesn't let us publish a transaction with no outputs). To work around these limitations, we
+    // start with a dummy output and later merge that dummy output with the optional change output added by bitcoind.
+    val txNotFunded = anchorTx.txInfo.tx.copy(txOut = TxOut(dustLimit, Script.pay2wpkh(PlaceHolderPubKey)) :: Nil)
+    // The anchor transaction is paying for the weight of the commitment transaction.
+    val anchorWeight = Seq(InputWeight(anchorTx.txInfo.input.outPoint, anchorInputWeight + commitTx.weight()))
+    bitcoinClient.fundTransaction(txNotFunded, FundTransactionOptions(targetFeerate, lockUtxos = true, inputWeights = anchorWeight)).flatMap(fundTxResponse => {
       // We merge the outputs if there's more than one.
       fundTxResponse.changePosition match {
         case Some(changePos) =>
           val changeOutput = fundTxResponse.tx.txOut(changePos)
-          val txSingleOutput = fundTxResponse.tx.copy(txOut = Seq(changeOutput.copy(amount = changeOutput.amount + dummyChangeAmount)))
-          Future.successful(fundTxResponse.copy(tx = txSingleOutput))
+          val txSingleOutput = fundTxResponse.tx.copy(txOut = Seq(changeOutput))
+          // We ask bitcoind to sign the wallet inputs to learn their final weight and adjust the change amount.
+          bitcoinClient.signTransaction(txSingleOutput, allowIncomplete = true).map(signTxResponse => {
+            val dummySignedTx = addSigs(anchorTx.updateTx(signTxResponse.tx).txInfo, PlaceHolderSig)
+            val packageWeight = commitTx.weight() + dummySignedTx.tx.weight()
+            val anchorTxFee = weight2fee(targetFeerate, packageWeight) - weight2fee(commitments.localCommit.spec.commitTxFeerate, commitTx.weight())
```

Comment:
Could `anchorTxFee` be < 0 (because `targetFeerate` is much lower than `commitTxFeerate` for example) ?

Reply:
I don't think it can, because we always have `targetFeerate > commitTxFeerate`, otherwise we don't even claim the anchor output. This is done at the very beginning of the `fund` method.

Reply:
I can add a `require` here to make this explicit?

Reply:
You're right it's not needed, this method is private and would not be called if the new target fee rate was lower the the current one.